revolv/payments/signals.py

```diff
@@ -0,0 +1,146 @@
+
+
+@receiver(signals.pre_init, sender=AdminRepayment)
+def pre_init_admin_repayment(**kwargs):
+    """
+    Make sure that related project is indeed complete, else throw a
+    ProjectNotCompleteException and to disallow instantiation of an invalid
+    AdminRepayment.
+    """
+    init_kwargs = kwargs.get('kwargs')
+    if not init_kwargs:
+        raise NotEnoughFundingException()
+    project = init_kwargs.get('project')
+    if not project.is_completed:
+        raise ProjectNotCompleteException()
+
+
+@receiver(signals.post_save, sender=AdminRepayment)
+def post_save_admin_repayment(**kwargs):
+    """
+    When an AdminRepayment is saved, a Repayment is generated for all donors to
+    a project, each weighed by that donor's proportion of the contribution to
+    the project.
+    """
+    instance = kwargs.get('instance')
+    for donor in instance.project.donors.all():
+        repayment = Repayment(user=donor,
+                              project=instance.project,
+                              admin_repayment=instance,
+                              amount=(instance.project.proportion_donated(donor) * instance.amount),
+                              )
+        repayment.save()
+
+
+@receiver(signals.pre_init, sender=AdminReinvestment)
+def pre_init_admin_reinvestment(**kwargs):
+    """
+    Raises a NotEnoughFundingException before __init__ if there are not enough
+    funds for this AdminReinvestment.
+    """
+    init_kwargs = kwargs.get('kwargs')
+    if not init_kwargs:
+        raise NotEnoughFundingException()
+    invest_amount = init_kwargs.get('amount') or 0.0
+
+    global_repay_amount = AdminRepayment.objects.aggregate(
+        Sum('amount')
+    )['amount__sum'] or 0.0
+    global_reinvest_amount = AdminReinvestment.objects.aggregate(
+        Sum('amount')
+    )['amount__sum'] or 0.0
+    global_reinvest_pool = global_repay_amount - global_reinvest_amount
+    if global_reinvest_pool - invest_amount < 0.0:
+        raise NotEnoughFundingException()
+
+
+@receiver(signals.post_save, sender=AdminReinvestment)
+def post_save_admin_reinvestment(**kwargs):
+    """
+    When an AdminReinvestment is saved, we pool as many donors as we need to
+    fund the reinvestment, prioritizing users that have a preference for the
+    Category of the project begin invested into. We only consider users that
+    have a non-zero pool of investable money.
+
+    !!! TODO: actually prioritize by Category
+    """
+    instance = kwargs.get('instance')
+    total_left = instance.amount
+    pending_reinvestors = []
+    for user in RevolvUserProfile.objects.filter(reinvest_pool__gt=0.0):
+        total_left -= user.reinvest_pool
+        reinvest_amount = user.reinvest_pool + min(0.0, total_left)
+        pending_reinvestors.append((user, reinvest_amount))
+        if total_left <= 0.0:
+            break
+    for (user, amount) in pending_reinvestors:
+        reinvestment = Payment(user=user,
```

Comment:
I feel like a comment here reminding the reader that the `user.reinvest_pool` will be decremented by the post_save signal on the Payment would be helpful